build.sc

```diff
@@ -0,0 +1,220 @@
+    override def description: String =
+      "Abstractions for contravariant tracing."
+  }
+
+  /** Additional crypto utilities such as threshold signature. */
+  object crypto extends SubModule with Publishing {
+    override def description: String =
+      "Cryptographic primitives to support HotStuff and BFT proof verification."
+
+    // TODO: Use crypto library from Mantis.
+    object test extends TestModule
+  }
+
+  /** Generic HotStuff BFT library. */
+  object hotstuff extends SubModule {
+
+    /** Pure consensus models. */
+    object consensus extends SubModule {
+      object test extends TestModule
+    }
+
+    /** Expose forensics events via tracing. */
+    object forensics extends SubModule
+
+    /** Implements peer-to-peer communication, state and block synchronisation.
+      *
+      * Includes the remote communication protocol messages and networking.
+      */
+    object service extends SubModule {
+      override def moduleDeps: Seq[JavaModule] =
+        Seq(storage, tracing, crypto, hotstuff.consensus, hotstuff.forensics)
+
+      override def ivyDeps = super.ivyDeps() ++ Agg(
+        ivy"io.iohk::scalanet:${VersionOf.scalanet}"
+      )
+
+      object test extends TestModule
+    }
+  }
+
+  /** Components realising the checkpointing functionality using HotStuff. */
+  object checkpointing extends SubModule {
+
+    /** Library to be included on the PoW side to talk to the checkpointing service.
+      *
+      * Includes the certificate models, the local communication protocol messages and networking.
+      */
+    object interpreter extends SubModule with Publishing {
+      override def description: String =
+        "Components to implement a PoW side checkpointing interpreter."
+
+      override def ivyDeps = Agg(
+        ivy"io.iohk::scalanet:${VersionOf.scalanet}"
+      )
+
+      override def moduleDeps: Seq[PublishModule] =
+        Seq(tracing, crypto)
+    }
+
+    /** Implements the checkpointing functionality and the ledger rules.
+      *
+      * If it was published, it could be directly included in the checkpoint assisted blockchain application,
+      * so the service and the interpreter can share data in memory.
+      */
+    object service extends SubModule {
+      override def moduleDeps: Seq[JavaModule] =
+        Seq(tracing, hotstuff.service, checkpointing.interpreter)
+
+      object test extends TestModule
+    }
+
+    /** Executable application for running HotStuff and checkpointing as a stand-alone process,
+      * communicating with the interpreter over TCP.
+      */
+    object app extends SubModule {
+      override def moduleDeps: Seq[JavaModule] =
+        Seq(hotstuff.service, checkpointing.service, rocksdb, logging, metrics)
+
+      override def ivyDeps = super.ivyDeps() ++ Agg(
+        ivy"com.typesafe:config:${VersionOf.config}",
+        ivy"ch.qos.logback:logback-classic:${VersionOf.logback}",
+        ivy"io.iohk::scalanet-discovery:${VersionOf.scalanet}",
```

Comment:
why the discovery is needed here ?

Reply:
The current checkpointing works on a fixed topology file. I read in the analysis that this caused some problems with Nomad, where you generally don't know on what machine deployments will end up on. With discovery as long as at least 1 machine's IP is known, the rest can be configured to bootstrap from it.

I also thought it would be a good feature towards the end to allow the federation to move their machines to different IPs. 

We could enhance discovery to be able to restrict the participants by the keys they use, so a random person cannot join the discovery protocol to learn where every federation member is.

Reply:
on one hand it seems reasonable, on other it is just adding another vector of attack on checkpointing nodes (exposed open udp port, whole kademlia protocol running in the background etc.) 

Fixed topology seems just inherently safer and with less moving parts. 

Lets see how it will play out.

Reply:
Yeah, it's one of the last things I thought we should do, after everything works with a fixed topology.